### PR TITLE
Issue #1192 Add test for version command

### PR DESCRIPTION
**Pull request by agajdosi**
Also adds support for regexp matching of stdout, stderr etc in format:
```
Then stdout should match "^minishift v\d\.\d\.\d\+[\w\d]{8}\n$"
And stderr should not match
"""
^minishift v\d\.\d\.\d\+[\w\d]{8}\n$
"""
```

### CODE REVIEW

**Comment by agajdosi** [comment]
Tagged WIP as there is currently an issue #1204 with `minishift version`. After fix I will edit the matching step and this will be ready.

I am adding this test as separate feature file as I would like to cover all commands in future. This is quite short and could be part of basic.feature, but as part of bigger plan I think it could make sense.

**Comment by agajdosi** [comment]
Or maybe we can go for behaviour - for example behaviour of minishift / its feature that it can show version information about itself, oc, docker etc.

File: test/integration/features/cmd/version.feature [hunk 1]
```diff
@@ -0,0 +1,7 @@
+@minishift-only @command
+Feature: Version command
+Command "minishift version" shows version of minishift binary.
+
+  Scenario: Setup-cdk installs and enables specific addons
```
Comment (hferentschik) [inline test/integration/features/cmd/version.feature]: I guess this is a cut&paste error. Let's have a correct scenario description here.

File: test/integration/features/cmd/version.feature [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+      When executing "minishift version" succeeds
+      Then stdout should match "^minishift v\d\.\d\.\d\+[\w\d]{7}\n$"
```
Comment (hferentschik) [inline test/integration/features/cmd/version.feature]: FYI - https://github.com/minishift/minishift/blob/d3c8dd05c51357651337cb7011a909b1c3452b2f/cmd/minishift/cmd/version_test.go#L34-L34

File: test/integration/features/cmd/version.feature [hunk 3]
```diff
@@ -0,0 +1,7 @@
+@minishift-only @command
```
Comment (hferentschik) [inline test/integration/features/cmd/version.feature]: why @command and why in a sub-directory _cmd_? _basic_ and _coolstore_ are also running a "bunch" of commands. Is the plan to create a feature for each command individually? The problem I see here is that a lot of commands will require a running Minishift instance. Provisioning a Minishift instance per command will take probably too much time. The version command is a bit special here since it requires no running VM. 

Either way, unless we have a concrete plan with the _cmd_ sub-directory I would argue that there is no need yet to create new directories.

File: test/integration/integration_test.go [hunk 4]
```diff
@@ -279,31 +288,57 @@ func checkConfigKey(configPath string, condition string, keyPath string) error {
 
 func compareExpectedWithActualContains(expected string, actual string) error {
 	if !strings.Contains(actual, expected) {
-		return fmt.Errorf("Output did not match. Expected: %s, Actual: %s", expected, actual)
+		return fmt.Errorf("Output did not match. Expected: '%s', Actual: '%s'", expected, actual)
```
Comment (hferentschik) [inline test/integration/integration_test.go]: Personally I find _errors.New_ more concise than _fmt.Errorf_, but the end result is probably the same.

+1 for the quotes around '%s' though. I had more than one occasion where having quotes helps to distinguish cases where I was not sure whether something was not working in general or whether the actual string was empty.

File: test/integration/integration_test.go [hunk 5]
```diff
@@ -394,6 +429,22 @@ func variableShouldNotBeEmpty(variableName string) error {
...
+func commandReturnShouldMatch(commandField string, expected string) error {
+	return compareExpectedWithActualMatches(expected, selectFieldFromLastOutput(commandField))
+}
+
+func commandReturnShouldNotMatch(commandField string, notexpected string) error {
+	return compareExpectedWithActualNotMatches(notexpected, selectFieldFromLastOutput(commandField))
+}
+
+func commandReturnShouldMatchContent(commandField string, expected *gherkin.DocString) error {
+	return compareExpectedWithActualMatches(expected.Content, selectFieldFromLastOutput(commandField))
```
Comment (hferentschik) [inline test/integration/integration_test.go]: Is the content version using regexp confirmed to work? Does creating a regular expression across multiple new lines work? Aka, does something like this work:

     Then stdout should contain
     """
     - anyuid         :  .*   P(0)
     - admin-user     :  .*  P(0)
     """

In some languages this would not work out of the box. Not sure atm how Go's regexp support handles this.

File: test/integration/features/cmd/version.feature (hunk 3 above)
Comment (gbraad) [inline test/integration/features/cmd/version.feature]: I suggest to make it part of the naming, instead of a subfolder. Folders can always be introduced at a later stage

File: test/integration/integration_test.go [hunk 6]
```diff
@@ -161,6 +161,15 @@ func FeatureContext(s *godog.Suite) {
 	s.Step(`([^"]*) should be valid ([^"]*)$`,
 		shouldBeInValidFormat)
 
+	s.Step(`([^"]*) should match "([^"]*)"$`,
```
Comment (gbraad) [inline test/integration/integration_test.go]: add comment to indicate these are used with regex, as term 'match' is very generic and can be mistaken as equals.
or include 'expression' in the string or function name (I prefer function name)

File: test/integration/integration_test.go [hunk 7]
```diff
@@ -279,31 +288,57 @@ func checkConfigKey(configPath string, condition string, keyPath string) error {
...
+	}
+
+	if !expectedRegexp.MatchString(actual) {
+		return fmt.Errorf("Output did not match. Expected: '%s', Actual: '%s'", expected, actual)
+	}
+
+	return nil
+}
+
+func compareExpectedWithActualNotMatches(notexpected string, actual string) error {
```
Comment (gbraad) [inline test/integration/integration_test.go]: Duplication of code... the difference between `!expectedRegexp` and `notexpectedRegexp` is so minimal, so please create a function that can be reused:

  * it compiles the regex
  * it compares

In that case `compareExpectedWithActualMatches` and `compareExpectedWithActualNotMatches` re-use functionality and become much shorter.

File: test/integration/features/cmd/version.feature (hunk 3 above)
Comment (agajdosi) [inline test/integration/features/cmd/version.feature]: > Is the plan to create a feature for each command individually?

@hferentschik Plan is to create features which will focus more on commands and edge cases of their usage than on some kind of story as coolstore and basic do. However for version command there is not much to investigate.

> The problem I see here is that a lot of commands will require a running Minishift instance.

Yes, this bothers me, version is specific and can be alone in small feature file, for other commands there will be need to group them I think, take two or three commands which are not interfering and group them. Or maybe we could go for one command per feature to bring clarity - overall it would take a lot of time but still it would be faster than manual testing we are doing now. Would like to have 3.1.0 as last release where majority of testing did persons.

@gbraad +1 going to do the renaming.

File: test/integration/features/cmd/version.feature (hunk 1 above)
Comment (agajdosi) [inline test/integration/features/cmd/version.feature]: @hferentschik Yes, thanks for pointing out. So small file and even though I missed that...

File: test/integration/integration_test.go (hunk 5 above)
Comment (agajdosi) [inline test/integration/integration_test.go]: @hferentschik Yes, tried it and fortunately it seems that it works in golang just fine, have not find a problem yet, tested with:

```gherkin
  Scenario: Test
    When executing "minishift addons list"
    Then stdout should match "^- .* : .* P\(0\)\n- .* : .* P\(0\)\n- .* : .* P\(0\)\n$"
    Then stdout should match
    """
    ^- .* : .* P\(0\)
    - .* : .* P\(0\)
    - .* : .* P\(0\)\n$
    """
``` 

Test passes.

File: test/integration/integration_test.go (hunk 7 above)
Comment (agajdosi) [inline test/integration/integration_test.go]: @gbraad Thank for feedback, going to implement it. Also thinking if I can take code from regexp matching for config files, will try it.

File: test/integration/integration_test.go (hunk 4 above)
Comment (agajdosi) [inline test/integration/integration_test.go]: @hferentschik Glad you like '%s'. I looked onto errors.New vs fmt.Errorf topic, if I understand it correctly errors.New does not support formatting, but if you do not need it then it saves one extra call of fmt.Sprintf(), fmt.Errorf supports formatting, which is needed here. Went through integration_test.go, util.go and minishift.go in test folder and found one fmt.Errorf() call, which do not use formatting and can be turned into more effective errors.New(). Going to fix it.

File: test/integration/features/cmd/version.feature (hunk 2 above)
Comment (agajdosi) [inline test/integration/features/cmd/version.feature]: @hferentschik Useful, thanks. However I am not sure why there is `{7-8}`, should not --short flag give seven characters as sha result?

File: test/integration/integration_test.go (hunk 4 above)
Comment (gbraad) [inline test/integration/integration_test.go]: I do not think it is necessary to prematurely optimize. this one extra call internally is not going to kill us, and especially not in a test situation. consistency and readability are more important.

Besides, most of the optimizations are done by the compiler... never be the compiler yourself unless performance is really an issue.

Comment (agajdosi) [inline test/integration/integration_test.go]: @gbraad Consistency, we have circa 7 errors now which do not need formatting and these are using errors.New(), just one has fmt.Errorf().

Comment (gbraad) [inline test/integration/integration_test.go]: Original tests had 3x errors.New(), I do not see a need to modify them.

File: test/integration/minishift.go [hunk 8]
```diff
@@ -214,7 +214,7 @@ func (m *Minishift) rolloutServicesSuccessfully(servicesToCheck string) error {
 	}
 
 	if !success {
-		return fmt.Errorf("Not all successfully rolled out")
+		return errors.New("Not all successfully rolled out")
```
Comment (gbraad) [inline test/integration/minishift.go]: Eh, why?

Comment (gbraad) [inline test/integration/minishift.go]: This should actually be changed to keep records of the returned value to determine which service did NOT successfully deploy.

Comment (agajdosi) [inline test/integration/minishift.go]: Ack

File: test/integration/integration_test.go (hunk 4 above)
Comment (hferentschik) [inline test/integration/integration_test.go]: I agree that there is no need for optimisation. I also agree that we don't really have to change it. I just said, _errors.New_ communicates better the intend imo. Even w/o knowing anything about Go, I can understand what _errors.New_ does. _fmt.Errorf_ is less obvious imo. Personally I had to look up the implementation/docs the first time I saw this function called. 

Also, in the production code we definitely use more _errors.New_. I have not seen many (if any)  _fmt.Errorf_.

Either way, happy to keep it.

**Comment by gbraad** [comment]
I think I started to use for.Errorf as the godog uses this AFAIR. ;-P

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

File: test/integration/integration_test.go (hunk 4 above)
Comment (agajdosi) [inline test/integration/integration_test.go]: > I just said, errors.New communicates better the intend imo.

Agree, but if I need to use `%s`, then it would look like this:
```
errors.New(fmt.Sprintf("This happened: %s, just terrible!", err))
```

**Comment by hferentschik** [comment]
Rebased and merged. Added a commit to make sure this test is run on CentOS CI.
